rust/operator-binary/src/spark_controller.rs

```diff
@@ -0,0 +1,968 @@
+                    PropertyNameKind::Env,
+                ],
+                masters.clone().erase(),
+            ),
+        );
+    }
+    if let Some(workers) = &resource.spec.workers {
+        result.insert(
+            SparkRole::Worker.to_string(),
+            (
+                vec![
+                    PropertyNameKind::File(SPARK_ENV_SH.to_string()),
+                    PropertyNameKind::File(SPARK_DEFAULTS_CONF.to_string()),
+                    PropertyNameKind::File(SPARK_METRICS_PROPERTIES.to_string()),
+                    PropertyNameKind::Env,
+                ],
+                workers.clone().erase(),
+            ),
+        );
+    }
+    if let Some(history_servers) = &resource.spec.history_servers {
+        result.insert(
+            SparkRole::HistoryServer.to_string(),
+            (
+                vec![
+                    PropertyNameKind::File(SPARK_ENV_SH.to_string()),
+                    PropertyNameKind::File(SPARK_DEFAULTS_CONF.to_string()),
+                    PropertyNameKind::File(SPARK_METRICS_PROPERTIES.to_string()),
+                    PropertyNameKind::Env,
+                ],
+                history_servers.clone().erase(),
+            ),
+        );
+    }
+    result
+}
+
+/// Build [`ContainerPort`]s for the given rolegroup.
+///
+/// # Arguments
+/// * `_sc`              - The cluster resource object.
+/// * `rolegroup`        - The rolegroup for which to extract the pods.
+/// * `rolegroup_config` - The validated configuration for the rolegroup.
+///
+fn build_container_ports(
+    _sc: &SparkCluster,
+    rolegroup: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Vec<ContainerPort> {
+    build_ports(_sc, rolegroup, rolegroup_config)
+        .iter()
+        .map(|(name, value)| ContainerPort {
+            name: Some(name.clone()),
+            container_port: *value,
+            protocol: Some("TCP".to_string()),
+            ..ContainerPort::default()
+        })
+        .collect()
+}
+
+/// Extract all named ports from the given validated configuration.
+///
+/// # Arguments
+/// * `_sc`              - The cluster resource object.
+/// * `rolegroup`        - The rolegroup for which to extract the pods.
+/// * `rolegroup_config` - The validated configuration for the rolegroup.
+///
+fn build_ports(
+    _sc: &SparkCluster,
+    rolegroup: &RoleGroupRef<SparkCluster>,
+    rolegroup_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Vec<(String, i32)> {
+    match serde_yaml::from_str(&rolegroup.role).unwrap() {
+        SparkRole::Master => vec![
+            (
+                PORT_NAME_MASTER.to_string(),
+                rolegroup_config
+                    .get(&PropertyNameKind::File(String::from(SPARK_ENV_SH)))
+                    .and_then(|c| c.get(SPARK_ENV_MASTER_PORT))
+                    .unwrap_or(&String::from("7077"))
+                    .parse()
+                    .unwrap(),
```

Comment:
I know this should be covered via product-config. What about "bad" overrides?
I would not unwrap() but use "?" and let `build_ports` return a result.

Reply:
Done.